python/ray/train/base_trainer.py

```diff
@@ -182,6 +182,8 @@ def __init__(
 
         self._validate_attributes()
 
+        air_usage.tag_air_trainer(self)
```

Comment:
If this is in init, it'll get called a bunch of times. Will get converted to trainable, then another trainer init will happen. So, at least 2x. If running with Tune, `# of trials x 2` times.

Maybe this is fine - since we'll just keep overwriting the key with the same value?

Some ideas:
1. Move to `as_trainable()`. This should only get called once for both Tuner and Trainer.
2. Move to `trainer.fit()` (I prefer this). If moving it to `trainer.fit()`, that wouldn't catch the `Tuner(Trainer)` case. Maybe these two cases should be tagged separately though. From the data, we can still aggregate the total usage of `XTrainer` by adding up `XTrainer` only usage + `XTrainer` + Tuner usage.

Reply:
Re

> Move to trainer.fit() (I prefer this). If moving it to trainer.fit(), that wouldn't catch the Tuner(Trainer) case. Maybe these two cases should be tagged separately though. From the data, we can still aggregate the total usage of XTrainer by adding up XTrainer only usage + XTrainer + Tuner usage.

We actually also want to capture that together. Think of this metric as something reflecting AIR trainer usage in general, not necessarily in training.

> If this is in init, it'll get called a bunch of times. Will get converted to trainable, then another trainer init will happen. So, at least 2x. If running with Tune, # of trials x 2 times.

I don't quite follow why "then another trainer init will happen". But in general, I get your concerns. Just to throw some context here, telemetry is not perfect right now (not operating at the granularity that we want) and given its implementation limitation, the team (@gjoliver , @matthewdeng and me) has decided to go down simpler path now.

Reply:
@xwjiang2010 roughly, how many tag calls will be made? I want to make sure it is not too crazy and overload the GCS.

Reply:
worst case, it's on the order of the number of trials in the cluster at a time. Hopefully it won't be too big.
Does telemetry have some sort of rate limiting mechanism built in?
Just FYI, RLlib is basically doing the same thing here: https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/algorithms/algorithm.py?L3014:9

We are not worse than that case. (same order of magnitude)
If you are concerned, we can also do a condition check if this is on head node, or even driver process (if you can point me to how to do that)!